src/stream/src/executor/backfill/cdc/state.rs

```diff
@@ -12,71 +12,165 @@
         match self {
-            CdcBackfillStateImpl::Undefined => Ok(false),
-            CdcBackfillStateImpl::SingleTable(state) => state.check_finished().await,
+            CdcBackfillStateImpl::Undefined => Ok(CdcStateItem::default()),
+            CdcBackfillStateImpl::SingleTable(state) => state.restore_state().await,
+            CdcBackfillStateImpl::MultiTable(state) => state.restore_state().await,
         }
     }
 
-    pub async fn mutate_state(
-        &mut self,
-        last_binlog_offset: Option<CdcOffset>,
-    ) -> StreamExecutorResult<()> {
+    pub async fn mutate_state(&mut self, state_item: CdcStateItem) -> StreamExecutorResult<()> {
         match self {
             CdcBackfillStateImpl::Undefined => Ok(()),
-            CdcBackfillStateImpl::SingleTable(state) => {
-                state.mutate_state(last_binlog_offset).await
-            }
+            CdcBackfillStateImpl::SingleTable(state) => state.mutate_state(state_item).await,
+            CdcBackfillStateImpl::MultiTable(state) => state.mutate_state(state_item).await,
         }
     }
 
     pub async fn commit_state(&mut self, new_epoch: EpochPair) -> StreamExecutorResult<()> {
         match self {
             CdcBackfillStateImpl::Undefined => Ok(()),
             CdcBackfillStateImpl::SingleTable(state) => state.commit_state(new_epoch).await,
+            CdcBackfillStateImpl::MultiTable(state) => state.commit_state(new_epoch).await,
         }
     }
 }
 
 pub const BACKFILL_STATE_KEY_SUFFIX: &str = "_backfill";
 
-/// The state manager for single cdc table
-pub struct SingleTableState<S: StateStore> {
+pub struct MultiBackfillState<S: StateStore> {
+    /// Id of the backfilling table, will be the key of the state
+    split_id: String,
+    state_table: StateTable<S>,
+}
+
+impl<S: StateStore> MultiBackfillState<S> {
+    pub fn new(table_id: u32, state_table: StateTable<S>) -> Self {
+        Self {
+            split_id: table_id.to_string(),
+            state_table,
+        }
+    }
+
+    pub fn init_epoch(&mut self, epoch: EpochPair) {
+        self.state_table.init_epoch(epoch)
+    }
+
+    /// Restore the backfill state from storage
+    pub async fn restore_state(&self) -> StreamExecutorResult<CdcStateItem> {
+        let key = Some(self.split_id.clone());
+        match self
+            .state_table
+            .get_row(row::once(key.map(ScalarImpl::from)))
+            .await?
+        {
+            Some(row) => {
+                let finished = match row.datum_at(1) {
+                    Some(ScalarRefImpl::Bool(val)) => val,
+                    _ => return Err(anyhow!("invalid backfill state: backfill_finished").into()),
+                };
+
+                Ok(CdcStateItem {
+                    is_finished: finished,
+                    ..Default::default()
+                })
+            }
+            None => Ok(CdcStateItem::default()),
+        }
+    }
+
```

Comment:
Should there be some other function to regularly update the backfill state?
Seems only have functions which restore, and finish the backfill state.
While backfilling, I suppose on barrier we should checkpoint the state?